src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/Interlocked.cs

```diff
@@ -23,6 +23,25 @@ public static int CompareExchange(ref int location1, int value, int comparand)
 #endif
         }
 
+        // This is used internally by NativeAOT runtime in cases where having a managed
+        // ref to the location is unsafe (Ex: it is the syncblock of a pinned object).
+        // The intrinsic expansion for this overload is exactly the same as for the `ref int`
+        // variant and will go on the same path since expansion is triggered by the name and
+        // return type of the method.
+        // The important part is avoiding `ref *location` in the unexpanded scenario, like
+        // in a case when compiling the Debug flavor of the app.
+        [Intrinsic]
+        internal static unsafe int CompareExchange(int* location1, int value, int comparand)
+        {
+#if TARGET_X86 || TARGET_AMD64 || TARGET_ARM64 || TARGET_RISCV64
+            return CompareExchange(location1, value, comparand); // Must expand intrinsic
+#else
+            // test readability of the location
+            _ = *location1;
```

Comment:
```suggestion
            Debug.Assert(location1 != null);
```
I do not think we need this for the internal use. This can be assert